src/Xamarin.Android.Tools.Bytecode/Kotlin/KotlinFixups.cs

```diff
@@ -83,15 +83,33 @@ static void FixupClassVisibility (ClassFile klass, KotlinClass metadata)
 				Log.Debug ($"Kotlin: Hiding internal class {klass.ThisClass.Name.Value}");
 				klass.AccessFlags = SetVisibility (klass.AccessFlags, ClassAccessFlags.Private);
 
-				foreach (var ic in klass.InnerClasses) {
-					Log.Debug ($"Kotlin: Hiding nested internal type {ic.InnerClass.Name.Value}");
-					ic.InnerClassAccessFlags = SetVisibility (ic.InnerClassAccessFlags, ClassAccessFlags.Private);
-				}
+				foreach (var ic in klass.InnerClasses)
+					HideInternalInnerClass (ic, classes);
 
 				return;
 			}
 		}
 
+		static void HideInternalInnerClass (InnerClassInfo klass, IList<ClassFile> classes)
+		{
+			var existing = klass.InnerClassAccessFlags;
+
+			klass.InnerClassAccessFlags = SetVisibility (existing, ClassAccessFlags.Private);
+			Log.Debug ($"Kotlin: Hiding nested internal type {klass.InnerClass.Name.Value}");
+
+			// Setting the inner class access flags above doesn't technically do anything, because we output
+			// from the inner class's ClassFile, so we need to find that and set it there.
+			var kf = classes.FirstOrDefault (c => c.ThisClass.Name.Value == klass.InnerClass.Name.Value);
```

Comment:
I wonder if `classes` should become a [`KeyedCollection<string, ClassFile>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.keyedcollection-2.-ctor?view=net-5.0)?  I'm just worried that if `classes` gets "big", `.FirstOrDefault()` may become a time sink.

Reply:
Potentially, though this should be a rarely used code path.